x-pack/legacy/plugins/maps/public/layers/styles/vector/components/color/dynamic_color_form.js

```diff
@@ -7,56 +7,174 @@
-    onDynamicStyleChange(styleProperty.getStyleName(), { ...styleOptions, field });
+export class DynamicColorForm extends React.Component {
+  state = {
+    colorMapType: COLOR_MAP_TYPE.ORDINAL,
   };
 
-  const onColorChange = colorOptions => {
-    onDynamicStyleChange(styleProperty.getStyleName(), {
-      ...styleOptions,
-      ...colorOptions,
-    });
-  };
+  constructor() {
+    super();
+    this._isMounted = false;
+  }
 
-  let colorRampSelect;
-  if (styleOptions.field && styleOptions.field.name) {
-    colorRampSelect = (
-      <ColorRampSelect
-        onChange={onColorChange}
-        color={styleOptions.color}
-        customColorRamp={styleOptions.customColorRamp}
-        useCustomColorRamp={_.get(styleOptions, 'useCustomColorRamp', false)}
-        compressed
-      />
-    );
+  componentWillUnmount() {
+    this._isMounted = false;
+  }
+
+  componentDidMount() {
+    this._isMounted = true;
+    this._loadColorMapType();
+  }
+
+  componentDidUpdate() {
+    this._loadColorMapType();
+  }
+
+  async _loadColorMapType() {
+    const field = this.props.styleProperty.getField();
+    if (!field) {
+      return;
+    }
+    const dataType = await field.getDataType();
+    const colorMapType = CATEGORICAL_DATA_TYPES.includes(dataType)
+      ? COLOR_MAP_TYPE.CATEGORICAL
+      : COLOR_MAP_TYPE.ORDINAL;
+    if (this._isMounted && this.state.colorMapType !== colorMapType) {
+      this.setState({ colorMapType });
+    }
   }
 
-  return (
-    <Fragment>
-      <EuiFlexGroup gutterSize="none" justifyContent="flexEnd">
-        <EuiFlexItem grow={false}>{staticDynamicSelect}</EuiFlexItem>
-        <EuiFlexItem>
-          <FieldSelect
-            fields={fields}
-            selectedFieldName={_.get(styleOptions, 'field.name')}
-            onChange={onFieldChange}
+  _onFieldMetaOptionsChange = fieldMetaOptions => {
+    const options = {
+      ...this.props.styleProperty.getOptions(),
+      fieldMetaOptions,
+    };
+    this.props.onDynamicStyleChange(this.props.styleProperty.getStyleName(), options);
+  };
+
+  _getColorSelector() {
+    const { onDynamicStyleChange, styleProperty } = this.props;
+    const styleOptions = styleProperty.getOptions();
+
+    let colorSelect;
+    if (styleOptions.field && styleOptions.field.name) {
+      const onColorChange = colorOptions => {
+        const newColorOptions = {
+          type: colorOptions.type,
+          color: colorOptions.color,
+        };
+        if (colorOptions.type === COLOR_MAP_TYPE.ORDINAL) {
+          newColorOptions.useCustomColorRamp = colorOptions.useCustomColorMap;
+          newColorOptions.customColorRamp = colorOptions.customColorMap;
+        } else {
+          newColorOptions.useCustomColorPalette = colorOptions.useCustomColorMap;
+          newColorOptions.customColorPalette = colorOptions.customColorMap;
+        }
+
+        const oldStyleOptions = { ...styleOptions };
+
+        if (oldStyleOptions.type === !newColorOptions.type) {
```

Comment:
Are any of these deletes needed? Line 94 will override `type` from `newColorOptions`. Then since colorPalette and colorRamp do not interfere with each other, why delete the other? I think it would be nice to save state so users can switch back and forth between the two and keep there changes

Reply:
this is to avoid clutter in the descriptor of the saved object.